File: Python_Solutions/Matrix_Chain_Multiplication/Egg_Dropping/Egg_Dropping_Memoization.py
import sys
def eggDrop_Memoization(e,f):
    t = [[-1 for j in range(f+1)] for i in range(e+1)]
    if(f == 0 or f == 1):
        return f
    if( e == 1):
        return f
    ans = sys.maxsize

    if(t[e][f] != -1):
        return t[e][f]
    for k in range(1, f+1):
        ans = min(ans, 1 + max(
            eggDrop_Memoization(e - 1, k - 1),
            eggDrop_Memoization(e, f - k)
        ))
    t[e][f] = ans
    return t[e][f]

n = 3
f = 5
print(eggDrop_Memoization(3,5))

File: CPP_Solutions/Matrix_Chain_Multiplication/Egg_Dropping/Egg_Dropping_Memoization.cpp
#include <bits/stdc++.h>
#include<algorithm>
using namespace std;

int t[1001][1001];

int eggDrop_Memoization(int e, int f)
{
    if (f == 1 || f == 0)
        return f;
    if (e == 1)
        return f;
        
    if(t[e][f] != -1)
        return t[e][f];
 
    int ans = INT_MAX, temp;
    for (int k = 1; k <= f; k++) {
        temp = 1+ std::max(
            eggDrop_Memoization(e - 1, k - 1),
            eggDrop_Memoization(e, f - k)
        );
        ans = std::min(ans, temp);
    }
 
    return t[e][f] = ans;
}

int main()
{
    memset(t, -1, sizeof(t));
    int n = 3;
    int k = 5;
    cout << eggDrop_Memoization(n , k);
    return 0;
}